include/threadpool.h

```diff
@@ -0,0 +1,86 @@
+#include <iostream>
+#include <string>
+#include <thread>
+#include <mutex>
+#include <vector>
+#include <queue>
+#include <condition_variable>
+#include <functional>
+
+#ifndef SMARTREDIS_THREADPOOL_H
+namespace SmartRedis {
+
+/*!
+*   \brief  A thread pool for concurrent execution of parallel jobs
+*/
+class ThreadPool
+{
+  public:
+    /*!
+    *   \brief ThreadPool constructor
+    *   \param num_threads The number of threads to create in the pool,
+    *          or 0 to use one thread per hardware context
+    */
+    ThreadPool(unsigned int num_threads=0);
```

Comment:
What about changing the default to `1`? That would allow `num_threads=0` to be a special case where we should bypass creating a threadpool.

Reply:
For performance, we want to use as many threads as possible without inducing context switches.  The default, 0, does this.  I don't expect users to be manually setting the environment variable for thread count in normal circumstances